contracts/strategies/LPOptimiserStrategy.sol

```diff
@@ -0,0 +1,270 @@
+
+        VaultParams memory vaultParams = _vaultParams[index];
+        IVoltzVault vault = _vaults[index];
+
+        // 0. Get tickspacing from vamm
+        int24 tickSpacing = vault.vamm().tickSpacing();
+
+        // 1. Get the new tick lower
+        int256 deltaWad = int256(currentFixedRateWad) - vaultParams.sigmaWad;
+        int256 newFixedLowerWad;
+        if (deltaWad > MINIMUM_FIXED_RATE) {
+            // delta is greater than MINIMUM_FIXED_RATE (0.01) => choose delta
+            if (deltaWad < vaultParams.maxPossibleLowerBoundWad) {
+                newFixedLowerWad = deltaWad;
+            } else {
+                newFixedLowerWad = vaultParams.maxPossibleLowerBoundWad;
+            }
+        } else {
+            // delta is less than or equal to MINIMUM_FIXED_RATE (0.01) => choose MINIMUM_FIXED_RATE (0.01)
+            newFixedLowerWad = MINIMUM_FIXED_RATE;
+        }
+        // 2. Get the new tick upper
+        int256 newFixedUpperWad = newFixedLowerWad + 2 * vaultParams.sigmaWad;
+
+        // 3. Convert new fixed lower rate back to tick
+        int256 newTickLowerWad = convertFixedRateToTick(newFixedUpperWad);
+
+        // 4. Convert new fixed upper rate back to tick
+        int256 newTickUpperWad = convertFixedRateToTick(newFixedLowerWad);
+
+        // 5. Scale ticks from wad
+        int256 newTickLowerExact = newTickLowerWad / 1e18;
+        int256 newTickUpperExact = newTickUpperWad / 1e18;
+
+        // 6. The underlying Voltz VAMM accepts only ticks multiple of tickSpacing
+        // Hence, we get the nearest usable tick
+        newTickLower = nearestTickMultiple(int24(newTickLowerExact), tickSpacing);
+        newTickUpper = nearestTickMultiple(int24(newTickUpperExact), tickSpacing);
+
+        // Call to VoltzVault contract to update the position lower and upper ticks
+        vault.rebalance(IVoltzVault.TickRange(newTickLower, newTickUpper));
+
+        emit RebalancedTicks(vault, newTickLower, newTickUpper);
+        return (newTickLower, newTickUpper);
+    }
+
+    /// @notice This function grabs all funds from the buffer vault 
+    /// and distributed them to the voltz vaults according to their weights
+    function _distributeTokens() internal {
+        // 0. Set the local variables
+        IERC20Vault erc20Vault = _erc20Vault;
+        address[] memory tokens = _tokens;
+        uint256 totalWeight = _totalWeight;
+
+        uint256[] memory balances = new uint256[](1);
+        balances[0] = IERC20(tokens[0]).balanceOf(address(erc20Vault));
+
+        // 1. Distribute the funds
+        uint256[] memory vaultShare = new uint256[](1);
+    
+        for (uint256 i = 0; i < _vaults.length; i++) {
+            uint256 vaultWeight = _vaultParams[i].weight;
+
+            if (vaultWeight == 0) {
+                continue;
+            }
+
+            // The share of i-th is vaultWeight / sum(vaultParams.weight)
+            vaultShare[0] = FullMath.mulDiv(balances[0], vaultWeight, totalWeight); 
```

Comment:
Due to rounding here, a situation may arise when not all tokens have been distributed. A possible solution is to send all remaining tokens to the vault with the highest weight.

Reply:
That's right but the remaining will be pushed in the next deposit and so on.